pkg/groupmapper/groupmapper.go

```diff
@@ -0,0 +1,220 @@
+	}
+}
+
+func (m *UserGroupsMapper) UserFor(identityInfo authapi.UserIdentityInfo) (kuser.Info, error) {
+	userInfo, err := m.delegatedUserMapper.UserFor(identityInfo)
+	if err != nil {
+		return userInfo, err
+	}
+
+	identityGroups := sets.NewString(identityInfo.GetProviderGroups()...)
+	if err := m.processGroups(identityInfo.GetProviderName(), identityInfo.GetProviderPreferredUserName(), identityGroups); err != nil {
+		return nil, err
+	}
+
+	return &UserInfoGroupsWrapper{
+		userInfo:         userInfo,
+		additionalGroups: identityGroups,
+	}, nil
+}
+
+func (m *UserGroupsMapper) processGroups(idpName, username string, groups sets.String) error {
+	err := wait.PollImmediate(1*time.Second, 5*time.Second, func() (bool, error) {
+		return m.groupsSynced(), nil
+	})
+	if err != nil {
+		return err
+	}
+
+	cachedGroups, err := m.groupsCache.GroupsFor(username)
+	if err != nil {
+		return err
+	}
+
+	removeGroups, addGroups := groupsDiff(cachedGroups, groups)
+	for _, g := range removeGroups {
+		if err := m.removeUserFromGroup(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	for _, g := range addGroups {
+		if err := m.addUserToGroup(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *UserGroupsMapper) removeUserFromGroup(idpName, username, group string) error {
+	updatedGroup, err := m.groupsLister.Get(group)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if len(updatedGroup.Users) == 0 {
+		return nil
+	}
+
+	if len(updatedGroup.Users) == 1 && updatedGroup.Users[0] == username && updatedGroup.Annotations[groupGeneratedKey] == "true" {
+		return m.groupsClient.Delete(context.TODO(), group, metav1.DeleteOptions{})
+	}
+
+	// don't perform any actions on the group if it hasn't been synced for this IdP
+	if updatedGroup.Annotations[fmt.Sprintf(groupSyncedKeyFmt, idpName)] != "synced" {
+		return nil
+	}
+
+	// find the user and remove it from the slice
+	userIdx := -1
+	for i, groupUser := range updatedGroup.Users {
+		if groupUser == username {
+			userIdx = i
+			break
+		}
+	}
+
+	var newUsers []string
+	switch userIdx {
+	case -1:
+		return nil
+	case 0:
+		newUsers = updatedGroup.Users[1:]
+	default:
+		newUsers = append(updatedGroup.Users[0:userIdx], updatedGroup.Users[userIdx+1:]...)
+	}
+
+	updatedGroupCopy := updatedGroup.DeepCopy()
+	updatedGroupCopy.Users = newUsers
+
+	_, err = m.groupsClient.Update(context.TODO(), updatedGroupCopy, metav1.UpdateOptions{})
+	return err
+}
+
+func (m *UserGroupsMapper) addUserToGroup(idpName, username, group string) error {
+	updatedGroup, err := m.groupsLister.Get(group)
+	if errors.IsNotFound(err) {
+		_, err = m.groupsClient.Create(context.TODO(),
+			&userv1.Group{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: group,
+					Annotations: map[string]string{
+						fmt.Sprintf(groupSyncedKeyFmt, idpName): "synced",
+						groupGeneratedKey:                       "true",
```

Comment:
Nit:

Sorry, I just now realized this. Would it make sense to make `"synced"` and `"true"` constants?